### DISCUSSION TITLE
Recommended upgrade path from V1 to V2 for nested objects/arrays

### DISCUSSION THREAD

**Discussion by jpblair** [dynamoose/dynamoose#1389] [Q&A] [answered]
Hi,

We're currently running Dynamoose 1.x. Some of our models had fields defined as `Object`s nested schemas, which were being `JSON.stringify()`ed as described in #230:
```
document: {
  type: Object,
  website: {
    type: String
  }
  elements: [Object]
}
```
```
"document": {
  "S": "{\"type\":\"collection\",\"website\":\"foo\",\"elements\":[]}"
}
```

When I try to upgrade to Dynamoose 2.x, I'm getting a conflict because the existing type of `S` for the `document` field conflicts with the model declaration of `Object`. I was wondering what the recommended way is to navigate this breaking change? Will we need to migrate to a new field, or is there an easier option? Thanks, and apologies if this has already been discussed elsewhere.

**Comment by fishcharlie** [comment, marked as answer]
@jpblair Sadly, the only option is to write your own custom script get the item, parse the JSON into an object, and resave it to your table.

This is likely the biggest breaking change in v2, and the biggest hassle. But also having JSON be stored as strings just doesn't make sense anymore.

There are a few considerations you could consider to limit server downtime. For example, you could create a new table, copy your data over to that table (parsing the JSON in the process), then use DynamoDB Streams or something on your old table whenever a change is made to make the identical change in your new table. At that point you should have 2 identical tables, 1 that is JSON strings, and 1 that is parsed JSON. Then when you are ready, simply point DynamoDB at that new table, and delete the old one.

Another thing. v3 originally had some discussion around modifiers before type checking occurred. We still haven't settled on a pattern for this and exactly how it'll work, and right now it's looking unlikely it'll be part of the initial release of v3.0.0. But might be coming in a future version. This would allow you to easily parse your data before it gets sent through the type checker.

Finally, you **might** be able to use [`dynamoose.aws.converter.set(converter)`](https://dynamoosejs.com/guide/Dynamoose#dynamooseawsconvertersetconverter) to build a customer converter. This has a lot of risks tho. You don't really know what table this refers to, you might not be parsing the correct field here, and this is also changing a lot in v3. I **highly** recommend against this solution. I'd personally put in the effort to clean your data.

I'm sorry I can't provide something better for you. I know migrating from v1 to v2 is a struggle. I do promise tho v3 is a lot easier, and won't have these complications about having to modify your underlying data. v2 really set us up for success moving forward, and although these decisions were tough, Dynamoose is a lot more stable today than it was with v1.

Here to help if you have any further questions or if I can assist any further.
